src/crimson/os/seastore/seastore_types.h

```diff
@@ -1051,34 +1071,172 @@ class laddr_t {
+  loffset_t get_offset() const { return offset; }
+  laddr_t checked_to_laddr() const {
+    assert(offset == 0);
+    return base;
+  }
+
+  /// static helper functions could access both fields of laddr_t and loffset_t
+  static laddr_offset_t plus(const laddr_t &laddr, loffset_t offset) {
+    auto base = laddr;
+    base.value += offset >> laddr_t::UNIT_SHIFT;
+    assert(base.value >= laddr.value);
+    offset &= laddr_t::UNIT_MASK;
+    return laddr_offset_t(base, offset);
+  }
+  static laddr_offset_t minus(const laddr_t &laddr, loffset_t offset) {
+    auto base = laddr;
+    auto diff = (offset + laddr_t::UNIT_SIZE - 1) >> laddr_t::UNIT_SHIFT;
+    base.value -= diff;
+    assert(base.value <= laddr.value);
+    offset = (diff << laddr_t::UNIT_SHIFT) - offset;
+    return laddr_offset_t(base, offset);
+  }
+
+  auto operator<=>(const laddr_offset_t&) const = default;
+
+private:
+  laddr_t base;
+  loffset_t offset;
+};
+std::ostream &operator<<(std::ostream &, const laddr_offset_t &);
+
+/// ================ overloads for laddr_t/laddr_offset_t begin ================
+inline bool operator==(
+  const laddr_t &laddr,
+  const laddr_offset_t &laddr_offset) {
+  return laddr == laddr_offset.get_aligned_laddr()
+      && 0 == laddr_offset.get_offset();
+}
+inline bool operator==(
+  const laddr_offset_t &laddr_offset,
+  const laddr_t &laddr) {
+  return laddr_offset.get_aligned_laddr() == laddr
+      && laddr_offset.get_offset() == 0;
+}
+
+inline auto operator<=>(
+  const laddr_t &laddr,
+  const laddr_offset_t &laddr_offset) {
+  return laddr_offset_t(laddr, 0) <=> laddr_offset;
+}
+inline auto operator<=>(
+  const laddr_offset_t &laddr_offset,
+  const laddr_t &laddr) {
+  return laddr_offset <=> laddr_offset_t(laddr, 0);
+}
+
+inline laddr_offset_t operator+(
+  const laddr_t &laddr,
+  const loffset_t &offset) {
+  return laddr_offset_t::plus(laddr, offset);
+}
+inline laddr_offset_t operator+(
+  const laddr_offset_t &laddr_offset,
+  const loffset_t &offset) {
+  return laddr_offset_t::plus(
+    laddr_offset.get_aligned_laddr(),
+    laddr_offset.get_offset() + offset);
+}
+inline laddr_offset_t operator+(
+  const loffset_t &offset,
+  const laddr_t &laddr) {
+  return laddr + offset;
+}
+inline laddr_offset_t operator+(
+  const loffset_t &offset,
+  const laddr_offset_t &loffset) {
+  return loffset + offset;
+}
+
+inline laddr_offset_t operator-(
+  const laddr_t &laddr,
+  const loffset_t &offset) {
+  return laddr_offset_t::minus(laddr, offset);
+}
+inline laddr_offset_t operator-(
+  const laddr_offset_t &laddr_offset,
+  const loffset_t &offset) {
+  if (laddr_offset.get_offset() >= offset) {
+    return laddr_offset_t(
+      laddr_offset.get_aligned_laddr(),
+      laddr_offset.get_offset() - offset);
+  } else {
+    return laddr_offset_t::minus(
+      laddr_offset.get_aligned_laddr(),
+      offset - laddr_offset.get_offset());
+  }
+}
+inline laddr_t::Unsigned operator-(
```

Comment:
Let's return `laddr_offset_t` like the above operations, and introduce an explicit method `laddr_offset_t::to_byte_offset() -> loffset_t`.

This also allows to validate that overflow shouldn't happen during `to_byte_offset()`.

Same to the below 2 operations.

Reply:
Given that more semantic infomation, such as `pool`, `shard` and `crush`, will be included in `laddr_t` to make it static in the future, I'd prefer make `laddr_t` move away from the integral semantics so that `laddr_t::Unsigned` is returned here. We could do it in this PR, but the return type will be `laddr_t::Unsigned` in 128-bits laddr_t PR. What do you think about this approach?

Reply:
I'm still not sure which specific case(s) will make returning `laddr_t::Unsigned` better.

Seems to me returning a `laddr_offset_t` is still better than `laddr_t::Unsigned` in case of 128-bit because `laddr_offset_t` won't discard any information.

Specifically, for example `operator-()`, if we mistakenly minus two laddrs from different pool/shard/crush, or anything else due to unexpected bugs, requiring an explicit conversion `laddr_offset_t::to_byte_offset()` is safer.

Also, more complicated expressions such as `laddr_t +- laddr_t +- loffset_t +- laddr_offset_t`, promoting the type correctly is reasonable I think.

Reply:
If there's a concern that the object offset calculations with full `laddr_offset_t` isn't efficient, I think user can adjust the calculation order to involve laddr in the last step/only when necessary.

Reply:
> Seems to me returning a `laddr_offset_t` is still better than `laddr_t::Unsigned` in case of 128-bit because `laddr_offset_t` won't discard any information from calculations involving `laddr_offset_t`.

I think the result of minus between two laddrs should not contains the information from orignial laddr because `laddr_t` will have a method `get_prefix` to fetch the unique laddr prefix in an OSD, it's meaningless to get the prefix of difference of laddrs if its type is `laddr_t`.

 It's better to make the difference only represtent the byte distance of laddr in my opinion.

> Specifically, for example `operator-()`, if we mistakenly minus two laddrs from different pool/shard/crush, or anything else due to unexpected bugs, requiring an explicit conversion `laddr_offset_t::to_byte_offset()` is safer.

It's not convenient to assume the two laddrs always has the same prefix(pool/shard/crush and local object id in the future), the reason is the end of data content address might carry over to the next prefix, it's valid to subtract a laddr from this end to calculate the length between laddr and the end.

> Also, more complicated expressions such as `laddr_t +- laddr_t +- loffset_t +- laddr_offset_t`, promoting the type correctly is reasonable I think.

It's not allowed to add a laddr with another laddr. Only minus is allowed between laddr_t/laddr_offset_t.

Reply:
What about introducing new types:

```
struct block_alinged_integer;
struct block_alinged_integer_and_unaligned_offset;
struct laddr_t {
  laddr_t(block_alinged_integer);
};
struct laddr_offset_t {
  laddr_offset_t(block_alinged_integer_and_unaligned_offset);
};
block_alinged_integer operator-(laddr_t, laddr_t);
block_alinged_integer_and_unaligned_offset operator-(laddr_offset_t, laddr_offset_t);
```

The basic idea is the construction of `laddr_t` and `laddr_offset_t` from the result of minus must be explicit. And the difference of laddrs doesn't inherit the information of original laddr_t.

Reply:
Reasonable to me.

How about keeping `operator+`, which seems fine, and using `get_byte_distance()` to replace `operator-`.

It might be a good idea to introduce templates, for example:
```c++
template <std::unsigned_integral U>
U laddr_t::get_byte_distance(U);  // validate over/underflow shouldn't happen based on U
```

There are multiple types in SeaStore that can represent offsets for various purposes, probably this is the most safe/general way, what do you think?

(Not sure whether `segment_off_t` might be related here, which is a `std::integral` though)